### PR TITLE
kubectl-gadget: Multiple fixes around skipInfo logic 

### LINKED ISSUE #2773: "kubectl-gadget completion" errors out instead of just creating the completion file.
## Description

I am trying to package kubectl-gadget for openSUSE: I wanted to package the completion files.
With many other packages it is enough to just issue `something completion bash > some/file` and package the results.

The same happens with the installed package, this also errors out  but does not output the completion.

I also downloaded the latest release, it also errors out:

```
$ ./kubectl-gadget completion bash
FATA[0000] Searching for running Inspektor Gadget instances: Get "http://localhost:8080/apis/apps/v1/daemonsets?fieldSelector=metadata.name%3Dgadget&labelSelector=k8s-app%3Dgadget": dial tcp [::1]:8080: connect: connection refused 
$
```

## Impact

Only `gadgetctl` has bash completion in the package, but not `kubectl-gadget`

## Environment and steps to reproduce

1. Build the kubectl-gadget binary.
1. Call it and give it the `completion bash` command
1. It fails with the error given above

## Expected behavior

Completion should be printed out.

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Add the completion command to the list of commands we need to skip getting the deploy info for.

Fix the following issues:
- kubectl-gadget completion not working when IG is not deployed
- kubectl-gadget version not showing the client version when k8s is not reacheable

Fixes #2773

### Testing

I performed the following:

```bash 
$ mkdir /tmp/config

$ minikube start -p ig
$ cp $HOME/.kube/config /tmp/config/ig

$ minikube start -p non-ig
$ cp $HOME/.kube/config /tmp/config/non-ig


$ KUBECONFIG=/tmp/config/ig kubectl get nodes
NAME   STATUS   ROLES           AGE     VERSION
ig     Ready    control-plane   2m14s   v1.28.3

$ KUBECONFIG=/tmp/config/non-ig kubectl get nodes
NAME     STATUS   ROLES           AGE   VERSION
non-ig   Ready    control-plane   23s   v1.28.3

# Deploy ig to one of the clusters
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget deploy

$ KUBECONFIG=/tmp/config/ig kubectl get namespaces
NAME              STATUS   AGE
default           Active   3m54s
gadget            Active   58s
kube-node-lease   Active   3m54s
kube-public       Active   3m54s
kube-system       Active   3m55s


# kubectl-gadget

## without k8s 
$ KUBECONFIG=/dev/null ./kubectl-gadget
Collection of gadgets for Kubernetes developers
...

## without IG deployed 
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget
Collection of gadgets for Kubernetes developers
...

## with IG deployed 
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget
Collection of gadgets for Kubernetes developers
...

# kubectl-gadget -h 

## without k8s 
$ KUBECONFIG=/dev/null ./kubectl-gadget -h
WARN[0000] Failed to contact Inspektor Gadget. Help text could be incomplete.
WARN[0000] No running Inspektor Gadget instances found. Help text could be incomplete.
Collection of gadgets for Kubernetes developers
...

## without IG deployed 
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget -h
WARN[0000] No running Inspektor Gadget instances found. Help text could be incomplete.
Collection of gadgets for Kubernetes developers
...

## with IG deployed 
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget -h
Collection of gadgets for Kubernetes developers
...

# kubectl-gadget version 

## without k8s 
$ KUBECONFIG=/dev/null ./kubectl-gadget version
Client version: v0.27.0-200-gea640ffa8-dirty
Error: searching for running Inspektor Gadget instances: Get "http://localhost:8080/apis/apps/v1/daemonsets?fieldSelector=metadata.name%3Dgadget&labelSelector=k8s-app%3Dgadget": dial tcp 127.0.0.1:8080: connect: connection refused

## without IG deployed
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget version
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget version
Client version: v0.27.0-200-gea640ffa8-dirty
Server version: not installed

## with IG deployed
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget version
Client version: v0.27.0-200-gea640ffa8-dirty
Server version: v0.27.0-190-gb69483053-dirty

# kubectl-gadget completion bash 

## without k8s
$ KUBECONFIG=/dev/null ./kubectl-gadget completion bash
... (works)

## without IG deployed
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget completion bash
... (works as well)

## with IG deployed
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget completion bash
... (works as well)


# kubectl-gadget trace exec 

## without k8s
$ KUBECONFIG=/dev/null ./kubectl-gadget trace exec
FATA[0000] Searching for running Inspektor Gadget instances.Get "http://localhost:8080/apis/apps/v1/daemonsets?fieldSelector=metadata.name%3Dgadget&labelSelector=k8s-app%3Dgadget": dial tcp 127.0.0.1:8080: connect: connection refused

## without IG deployed
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget trace exec
FATA[0000] No running Inspektor Gadget instances found.

## with IG deployed
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget trace exec
K8S.NODE         K8S.NAMESPACE    K8S.POD          K8S.CONTAINER    PID       PPID     COMM     PCOMM    RET ARGS
(works...)


# kubectl-gadget trace exec -h

## without k8s
$ KUBECONFIG=/dev/null ./kubectl-gadget trace exec -h
WARN[0000] Failed to contact Inspektor Gadget. Help text could be incomplete.
WARN[0000] No running Inspektor Gadget instances found. Help text could be incomplete.
... 

## without IG deployed
$ KUBECONFIG=/tmp/config/non-ig ./kubectl-gadget trace exec -h
WARN[0000] No running Inspektor Gadget instances found. Help text could be incomplete.
... 

## with IG deployed
$ KUBECONFIG=/tmp/config/ig ./kubectl-gadget trace exec -h
Trace new processes
...

```

### CODE REVIEW

File: cmd/kubectl-gadget/main.go [hunk 1]
```diff
@@ -81,6 +82,7 @@ func main() {
 		isVersion = isVersion || arg == "version"
 		isDeployUndeploy = isDeployUndeploy || arg == "deploy" || arg == "undeploy"
 		isHelp = isHelp || arg == "--help" || arg == "-h"
+		isCompletion = isCompletion || arg == "completion"
```
[the remark this answers is not in the document]
Comment (burak-ok) [reply cmd/kubectl-gadget/main.go:85]: Some of them we could. Then there is e.g. `isVersion` which is needed separatly

**Review by burak-ok — APPROVED** [review]
LGTM

File: cmd/kubectl-gadget/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go]: I completely reworked this logic to avoid this. Can any of you folks test it again please? (Not only the completion, but other things to be sure I didn't break anything).

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go]: Would you mind describing a bit more the test you did? Did you have kubeconfig pointing to a cluster? Was Inspektor Gadget deployed?

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go]: I'm now in full refactoring more. See also https://github.com/inspektor-gadget/inspektor-gadget/issues/2773#issuecomment-2079394906. Let's me see if I can get something working.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go]: I updated the PR. I hope this time it works fine. Would you mind checking it again?

**Comment by mauriciovasquezbernal** [comment]
The CI is failing because the check-readme test is comparing the stderr lines as well. After expending some minutes trying to understand how to fix it in perl, I decided to open #2786 that fixes it by rewriting it in golang.

**Comment by burak-ok** [comment]
No full review yet. After I removed `info.json` and stopped my k8s cluster and tried to see the help, multiple warnings are printed. One seems to be a duplicate with a slightly different wording.
What kind of warnings do we want to print in this case, etc...

```
./kubectl-gadget help
WARN[0000] Failed to contact Inspektor Gadget. Help text could be incomplete. 
WARN[0000] No running Inspektor Gadget instances found. Help text could be incomplete. 
WARN[0000] could not load gadget info from remote: dialing random target: get gadget pods: getting pods: Get "http://localhost:8080/api/v1/namespaces/gadget/pods?labelSelector=k8s-app%3Dgadget": dial tcp 127.0.0.1:8080: connect: connection refused 
WARN[0000] Failed to load deploy info: opening info file: open /home/burak/.ig/info.json: no such file or directory 
Collection of gadgets for Kubernetes developers

Usage:
  kubectl-gadget [command]

Available Commands:
  advise      Recommend system configurations based on collected information
```

**Comment by mauriciovasquezbernal** [comment]
I updated the logic to print less warnings, I hope that helps.

File: cmd/kubectl-gadget/main.go [hunk 2]
```diff
@@ -104,38 +103,49 @@ func main() {
...
+				log.Warnf("Failed to contact Inspektor Gadget. Help text could be incomplete.")
+			} else {
+				log.Fatalf("Searching for running Inspektor Gadget instances.%s", err)
+			}
+		} else {
+			switch len(gadgetNamespaces) {
+			case 0:
+				if isHelp {
+					log.Warn("No running Inspektor Gadget instances found. Help text could be incomplete.")
+				} else {
```
Comment (burak-ok) [inline cmd/kubectl-gadget/main.go:119]:
```suggestion
				} else if !isVersion {
```

We need a separate switch for `isVersion` or we will not print out the client version if we didn't deploy IG yet.
Before:
```
./kubectl-gadget version
Client version: v0.27.0-202-g8d405757
Server version: not installed
```
After:
```
./kubectl-gadget version
FATA[0000] No running Inspektor Gadget instances found. 
```

Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go]: ohman, covering all corner cases in this PR is difficult. I adjusted the logic again. Let's hope this time it's right.

File: cmd/kubectl-gadget/main.go [hunk 3]
```diff
@@ -63,24 +59,20 @@ func main() {
...
+	}
+
+	// Need to loop through all arguments to skip flags...
 	for _, arg := range os.Args[1:] {
-		for _, skipCmd := range infoSkipCommands {
-			if arg == skipCmd {
-				skipInfo = true
-			}
+		switch arg {
+		case "completion", "deploy", "version":
```
Comment (burak-ok) [inline cmd/kubectl-gadget/main.go:73]:
```suggestion
		case "completion", "deploy", "version", "--help", "help", "-h":
```
There is now a difference in the help messages:
```
./kubectl-gadget                            
Collection of gadgets for Kubernetes developers

Usage:
  kubectl-gadget [command]
```
```
./kubectl-gadget --help
WARN[0000] No running Inspektor Gadget instances found. 
WARN[0000] Failed to load deploy info: loading gadget info from remote: dialing random target: get gadget pods: no gadget pods found in namespace "gadget". Is Inspektor Gadget deployed? 
Collection of gadgets for Kubernetes developers

Usage:
```

**Review by burak-ok — APPROVED** [review]
I'll approve. 
I think after that single change everything is fine. Way too many corner cases to think about. We should write tests for this instead to define how we want the CLI to behave

File: cmd/kubectl-gadget/main.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/main.go:73]: But then the deploy info won't be loaded and the help output will be incomplete:

```
...
Usage:
  kubectl-gadget [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  deploy      Deploy Inspektor Gadget on the cluster
  help        Help about any command
  undeploy    Undeploy Inspektor Gadget from cluster
  version     Show version
...
```

Comment (burak-ok) [reply cmd/kubectl-gadget/main.go:74]: Corner cases after corner cases. This whole logic is cursed
